diffy/plugins/diffy_osquery/plugin.py

```diff
@@ -32,30 +33,41 @@ def generate(self, incident: str, **kwargs) -> List[str]:
         logger.debug("Generating osquery payload.")
         session = Session()
 
-        # TODO check for existence before deployment
-        # we run with these commands with diffy credentials so as to not pollute the on-instance credentials
-        creds = session.get_credentials()
-        region = kwargs.get("region", CONFIG.get("DIFFY_PAYLOAD_OSQUERY_REGION"))
-        key = kwargs.get("key", CONFIG.get("DIFFY_PAYLOAD_OSQUERY_KEY"))
-
-        if not region:
-            raise BadArguments(
-                "DIFFY_PAYLOAD_OSQUERY_REGION required for use with OSQuery plugin."
-            )
-
-        if not key:
-            raise BadArguments(
-                "DIFFY_PAYLOAD_OSQUERY_KEY required for use with OSQuery plugin."
-            )
-
-        commands: List[str] = [
-            f"export AWS_ACCESS_KEY_ID={creds.access_key}",
-            f"export AWS_SECRET_ACCESS_KEY={creds.secret_key}",
-            f"export AWS_SESSION_TOKEN={creds.token}",
-            f"cd $(mktemp -d -t binaries-{incident}-`date +%s`-XXXXXX)",
-            f"aws s3 --region {region} cp s3://{key} ./latest.tar.bz2 --quiet",
-            "tar xvf latest.tar.bz2 &>/dev/null",
-        ]
+        # If osquery isn't present, obtain an osquery binary from S3.
+        if not which("osqueryi"):
+            # We run these commands with Diffy credentials so as to not pollute
+            # the on-instance credentials.
+            creds = session.get_credentials()
+            region = kwargs.get("region", CONFIG.get("DIFFY_PAYLOAD_OSQUERY_REGION"))
+            key = kwargs.get("key", CONFIG.get("DIFFY_PAYLOAD_OSQUERY_KEY"))
+
+            if not region:
+                raise BadArguments(
+                    "DIFFY_PAYLOAD_OSQUERY_REGION required for use with OSQuery plugin."
+                )
+
+            if not key:
+                raise BadArguments(
+                    "DIFFY_PAYLOAD_OSQUERY_KEY required for use with OSQuery plugin."
+                )
+
+            # If we've downloaded our own osquery collection binary, create a
+            # symbolic link, allowing us to use relative commands elsewhere.
+            commands: List[str] = [
+                f"export AWS_ACCESS_KEY_ID={creds.access_key}",
+                f"export AWS_SECRET_ACCESS_KEY={creds.secret_key}",
+                f"export AWS_SESSION_TOKEN={creds.token}",
+                f"cd $(mktemp -d -t binaries-{incident}-`date +%s`-XXXXXX)",
+                f"aws s3 --region {region} cp s3://{key} ./latest.tar.bz2 --quiet",
+                "tar xvf latest.tar.bz2 &>/dev/null",
+                "export PATH=${PATH}:${HOME}/.local/bin",
+                "mkdir -p ${HOME}/.local/bin",
+                "ln -s ./usr/bin/osqueryi ${HOME}/.local/bin/osqueryi",
+            ]
+        else:
+            commands: List[str] = [
+                f"cd $(mktemp -d -t binaries-{incident}-`date +%s`-XXXXXX)"
```

Comment:
Do we need to make this dir if we already have osquery available? Are there other binaries we always pull down? To put another way should we always pull down `latest.tar.bz2` (regardless of if we need osquery) and only do an additional linking step if it does not exist?

Reply:
Oh I mistook that line for concurrent use with _results_ or _artifacts_. Well, but if osquery isn't the payload, we should have a separate plugin for that, I think. `TrustedBinaryCollectionPayloadPlugin` or the like.

Reply:
+1 for `TrustedBinaryCollectionPayloadPlugin`

Reply:
Added #55 .